### PR TITLE
Feat data class into main 

### PR DESCRIPTION
**Pull request by RobinFeldmann**
### Changes
- Related Issues #27, #112, #71 
- Added BaseDataset Class and MultivariatTimeseriesDataset, example Dataset in "integration_tests/test_data.py"
- Added DatasetSpecification Dataclass
  - It contains neccessary dataset informations for Trainer to initialize Dataset
- Added Downloadable Base and DownloadableSciebo for multiple inheritance like
`class DownloadableDataset(BaseDataset, DownloadableSciebo)`
  - to use this: upload your dataset to "sciebo/datasets" as a csv, pandas or feather and set the correct sciebo-path in your DownloadableDataset
  -  example in "integration_tests/test_data.py"
- Added Trainer Baseclass, with validation, test, train, train_epoch, early stopping
  - Added ReconstructionAutoencoderTrainer 
    - Added Loss or Accuracy return to validate() and test()
- Added Tests
- Some refactoring

### CODE REVIEW

File: elasticai/explorer/hw_nas.py [hunk 1]
```diff
@@ -31,39 +38,46 @@ def evaluate_model(model: torch.nn.Module, device: str):
...
-        MNIST("data/mnist", download=True, train=False, transform=transf), batch_size=64
-    )
-    trainer = MLPTrainer(device, optimizer)
-
-    metric = {"default": 0, "accuracy": 0, "flops log10": math.log10(flops)}
-    for epoch in range(n_epochs):
-        trainer.train_epoch(model, train_loader, epoch)
 
-        metric["accuracy"] = trainer.test(model, test_loader)
+    trainer = trainer_class(device, optimizer, dataset_info)
```
Comment (chrisringhofer) [inline elasticai/explorer/hw_nas.py:42]: dataset_info ist nicht spezifisch genug und uneinheitlich -> dataset_spec

File: elasticai/explorer/training/data.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    @abstractmethod
+    def __len__(self) -> int:
+        pass
+
+    @abstractmethod
+    def __getitem__(self, idx) -> Any:
+        pass
+
+
+class MultivariatTimeseriesDataset(BaseDataset):
```
Comment (chrisringhofer) [inline elasticai/explorer/training/data.py:43]: Multivariate

File: elasticai/explorer/training/data.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        transform: Optional[Callable] = None,
+        target_transform: Optional[Callable] = None,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(root, train, transform, target_transform, *args, **kwargs)
+        self._setup_data()
+        self._setup_targets()
+
+        if train:
```
Comment (chrisringhofer) [inline elasticai/explorer/training/data.py:62]: Hier sollte nicht schon ein fixer train-test-split und fixer seed vorgegeben werden. Vorschlag 1: split in die Trainer-Klasse, dort konfigurieren. Vorschlag 2: extra Methode auf Objekt um Train-Test-Split zu generieren

File: elasticai/explorer/training/data.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            case _:
+                raise ValueError(f"Unsupported filetype: {file_path}")
+
+
+@dataclass
+class DatasetSpecification:
+    dataset_type: Type[MNIST] | Type[BaseDataset]
+    dataset_location: Path
+    transform: Compose | None = None
+    validation_split_ratio: List[float] = field(default_factory=lambda: [0.8, 0.2])
```
Comment (chrisringhofer) [inline elasticai/explorer/training/data.py:120]: Test-Split ebenfalls

File: elasticai/explorer/training/trainer.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+
+
+class Trainer(ABC):
+
+    def __init__(
+        self,
+        device: str,
+        optimizer: Optimizer,
+        dataset_spec: DatasetSpecification,
+        loss_fn: nn.modules.loss._Loss = nn.CrossEntropyLoss(),
```
Comment (chrisringhofer) [inline elasticai/explorer/training/trainer.py:19]: Type hint raus

File: elasticai/explorer/training/trainer.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+
+class MLPTrainer(Trainer):
+    """Trainer class for MLPs written in Pytorch."""
+
+    def __init__(
+        self,
+        device: str,
+        optimizer: Optimizer,
+        dataset_spec: DatasetSpecification,
+        loss_fn: nn.modules.loss._Loss = nn.CrossEntropyLoss(),
```
Comment (chrisringhofer) [inline elasticai/explorer/training/trainer.py:89]: Type hint raus

File: elasticai/explorer/training/trainer.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+        best_val_accuracy = 0.0
+        patience_counter = 0
+        best_model_state = model.state_dict()
+
+        for epoch in range(epochs):
+            self.logger.info(f"Epoch {epoch+1}/{epochs}")
+            self.train_epoch(model=model, epoch=epoch)
+            val_accuracy, _ = self.validate(model=model)
+
+            if val_accuracy > best_val_accuracy + self.min_delta:
```
Comment (chrisringhofer) [inline elasticai/explorer/training/trainer.py:123]: Besser immer aus loss beziehen

File: elasticai/explorer/training/trainer.py [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+        best_val_loss = 0.0
+        patience_counter = 0
+        best_model_state = model.state_dict()
+
+        for epoch in range(epochs):
+            self.logger.info(f"Epoch {epoch+1}/{epochs}")
+            self.train_epoch(model=model, epoch=epoch)
+            val_loss, _ = self.validate(model=model)
+
+            if val_loss < best_val_loss - self.min_delta:
```
[the remark this answers is not in the document]
Comment (chrisringhofer) [reply elasticai/explorer/training/trainer.py:262]: Zusätzlicher Vorschlag: evaluate() Methode zum @override bereitstellen, sodass train() bestenfalls immer gleich bleiben kann und die custom evaluate() Methode aufruft

File: elasticai/explorer/training/trainer.py [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+class Trainer(ABC):
+
+    def __init__(
+        self,
+        device: str,
+        optimizer: Optimizer,
+        dataset_spec: DatasetSpecification,
+        loss_fn: nn.modules.loss._Loss = nn.CrossEntropyLoss(),
+        batch_size: int = 64,
+        early_stopping: bool = True,
```
Comment (chrisringhofer) [inline elasticai/explorer/training/trainer.py:21]: in train() packen

File: mlp_for_pi.py
[the remark this answers is not in the document]
Comment (chrisringhofer) [reply mlp_for_pi.py:1]: Vielleicht "examples/" oder "app/" folder anlegen und dort reinschieben

**Review by chrisringhofer — CHANGES_REQUESTED** [review]
Größtenteils fein, einfach die Kommentare adressieren :)

File: elasticai/explorer/training/data.py (hunk 4 above)
Comment (RobinFeldmann) [reply elasticai/explorer/training/data.py]: Habe jetzt eine `test_train_val_ratio: List[float] = field(default_factory=lambda: [0.7, 0.1, 0.2]) ` eingeführt das macht das ganze auch übersichtlicher, da direkt die Verhältnisse für alle 3 Subsets klar sind.

File: elasticai/explorer/training/data.py (hunk 3 above)
Comment (RobinFeldmann) [reply elasticai/explorer/training/data.py]: Der Split liegt jetzt komplett im Trainer.

File: elasticai/explorer/utils/utils.py [hunk 10]
```diff
@@ -1,21 +1,22 @@
...
 from pathlib import Path
+from typing import Any
 
 import pandas
 import plotly.express as px
 from scipy.stats import kendalltau
 
 
-def compute_kendall(list_x: list[any], list_y: list[any]) -> float:
+def compute_kendall(list_x: list[Any], list_y: list[Any]) -> Any:
```
[the remark this answers is not in the document]
Comment (RobinFeldmann) [reply elasticai/explorer/utils/utils.py:11]: Hab es mal trotzdem schon gemacht.

File: mlp_for_pi.py
[the remark this answers is not in the document]
Comment (RobinFeldmann) [reply mlp_for_pi.py:1]: Gemacht, aber typisches Python + VSC Problem, man braucht workaround für Package import.

**Review by chrisringhofer — DISMISSED** [review]
Alles fein! :-)

Bei test_train_val_ratio könnte man überlegen, ob wir noch Verhalten definieren wollen für den Fall, dass die Ratios sich zu > 1.0 aufaddieren. https://docs.pytorch.org/docs/stable/data.html#torch.utils.data.random_split kümmert sich nur um den Fall, dass die Ratios sich zu < 1.0 aufaddieren. Sehe ich aber als absolut nicht wichtig an jetzt - liegt dann halt einfach in der Verantwortung der App-Entwickler, das sicherzustellen. Ist auch nicht zu viel verlangt ;-)

Von daher: alles akzeptiert, merge gern! (Merging ist offenbar noch blockiert)

**Comment by RobinFeldmann** [comment]
Ja, mussten noch alle Konversationen auf "Resolved" stellen. Wie abgesprochen merge ich aber noch nicht, da Optuna vorrang hat.

**Review by mokouMonday — APPROVED** [review]
